### PR TITLE
Sync: Propagate block announcement data

### PR DESCRIPTION
**Pull request by bkchr**
This pr adds a feature to the sync protocol to propagate the data that
we received alongside a block announcement. This is done by adding a
cache that caches the last X block announcement data where X is set to
the number of `in_peers` (giving every peer the chance to send us a
different block). This will be required by parachains to ensure that
even peers who are not connected to a collator receive the data
alongside the block announcement to properly validate it and request the
block.

### CODE REVIEW

File: client/network/src/protocol.rs [hunk 1]
```diff
@@ -1157,24 +1166,37 @@ impl<B: BlockT, H: ExHashT> Protocol<B, H> {
 		validation_result: sync::PollBlockAnnounceValidation<B::Header>,
 	) -> CustomMessageOutcome<B> {
 		let (header, is_best, who) = match validation_result {
-			sync::PollBlockAnnounceValidation::Nothing { is_best, who, header } => {
+			sync::PollBlockAnnounceValidation::Nothing { is_best, who, announce } => {
 				self.update_peer_info(&who);
 
+				if let Some(data) = announce.data {
+					if !data.is_empty() {
+						self.block_announce_data_cache.put(announce.header.hash(), data);
```
Comment (andresilva) [inline client/network/src/protocol.rs:1174]: In this case should we overwrite any existing block announce data that might already exist? My understanding is that we might not actually validate the block announce data when we get `PollBlockAnnounceValidation::Nothing`.

File: client/network/test/src/sync.rs [hunk 2]
```diff
@@ -842,3 +842,58 @@ fn sync_to_tip_when_we_sync_together_with_multiple_peers() {
 		net.block_until_idle();
 	}
 }
+
+/// Ensures that when we receive a block announcement with some data attached, that we propagate
+/// this data when reannouncing the block.
+#[test]
+fn block_announce_data_is_propagated() {
```
Comment (andresilva) [inline client/network/test/src/sync.rs:849]: Nice

File: client/network/test/src/sync.rs [hunk 3]
```diff
@@ -842,3 +842,58 @@ fn sync_to_tip_when_we_sync_together_with_multiple_peers() {
...
+			Poll::Ready(())
+		} else {
+			Poll::Pending
+		}
+	}));
+
+	let block_hash = net.peer(0).push_blocks_at_without_announcing(BlockId::Number(0), 1, true);
+	net.peer(0).announce_block(block_hash, Some(vec![137]));
+
+	while !net.peer(1).has_block(&block_hash) || !net.peer(2).has_block(&block_hash) {
```
Comment (andresilva) [inline client/network/test/src/sync.rs:896]: Shouldn't we be asserting that peer 2 got the block announcement with the announcement data attached as well?

**Review by andresilva — CHANGES_REQUESTED** [review]

File: client/network/test/src/sync.rs (hunk 3 above)
Comment (bkchr) [reply client/network/test/src/sync.rs:896]: It uses the same block announcement validator: https://github.com/paritytech/substrate/pull/7903/files/3cd67a501ca7305f441f7e1b6c9c55144ed1d878#diff-4f2a05557303da2df61eeb7972eb4b1e290db61cfce21f22868d0ce90c77df6eR878

This rejects any block announcement that isn't in the expected format.

File: client/network/src/protocol.rs (hunk 1 above)
Comment (bkchr) [reply client/network/src/protocol.rs]: Good point. Removed it.

Comment (bkchr) [reply client/network/src/protocol.rs:1176]: Okay readded it xD But changed the logic a little bit :P

**Review by andresilva — APPROVED** [review]

**Comment by arkpar** [comment]
Total number of peers may be significantly different than `in_peers` since there's also `out_peers` and also reserved peers and other priority groups on top of that. I'd use a hardcoded constant here or at least `in_peers + out_peers`

**Comment by bkchr** [comment]
> Total number of peers may be significantly different than `in_peers` since there's also `out_peers` and also reserved peers and other priority groups on top of that. I'd use a hardcoded constant here or at least `in_peers + out_peers`

I changed it to the later. However, this is only relevant for different blocks. If we see `in_peers + out_peers` different blocks floating around, we probably have other problems :D

**Review by arkpar — APPROVED** [review]

**Pull request merged**